server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala

```diff
@@ -0,0 +1,164 @@
+              case Some(response) => response
+              case None           => ServerResponse.notFound
+            }
+        }
+
+        val _ = runningFuture.transform {
+          case Success(serverResponse) =>
+            try {
+              serverResponse.body match {
+                case Some(function) => {
+                  val content = function(ctx)
+                  content match {
+                    case r: ReactiveWebSocketProcessorNettyResponseContent => {
+                      ctx
+                        .pipeline()
+                        .addAfter(
+                          "serverCodecHandler",
+                          "wsControlFrameHandler",
+                          new NettyControlFrameHandler(
+                            ignorePong = r.ignorePong,
+                            autoPongOnPing = r.autoPongOnPing,
+                            decodeCloseRequests = r.decodeCloseRequests
+                          )
+                        )
+                      r.autoPing.foreach { case (interval, pingMsg) =>
+                        ctx
+                          .pipeline()
+                          .addAfter("wsControlFrameHandler", "wsAutoPingHandler", new WebSocketAutoPingHandler(interval, pingMsg))
+                      }
+                      // Manually completing the promise, for some reason it won't be completed in writeAndFlush. We need its completion for NettyBodyListener to call back properly
+                      r.channelPromise.setSuccess()
+                      val _ = ctx.writeAndFlush(
+                        // Push a special message down the pipeline, it will be handled by HttpStreamsServerHandler
+                        // and from now on that handler will take control of the flow (our NettyServerHandler will not receive messages)
+                        new DefaultWebSocketHttpResponse(
+                          req.protocolVersion(),
+                          HttpResponseStatus.valueOf(200),
+                          r.processor, // the Processor (Pipe) created by Tapir interpreter will be used by HttpStreamsServerHandler
+                          new WebSocketServerHandshakerFactory(wsUrl(req), null, false)
+                        )
+                      )
+                    }
+                    case otherContent =>
+                      logger.error(s"Unexpected response content: $otherContent")
+                      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.INTERNAL_SERVER_ERROR)
+                      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+                      otherContent.channelPromise.setFailure(new IllegalStateException("Unexpected response content"))
+                      val _ = ctx.writeAndFlush(res)
+                  }
+                }
+                case None =>
+                  logger.error("Missing response body, expected WebSocketProcessorNettyResponseContent")
```

Comment:
is this really an error? that is, won't this occur when a client attempts to upgrade a "normal" endpoint to a websocket? (that would be client's error, but nothing to fix on the server-side)

Reply:
yes, it's definitely a client error, not a server one, so 4xx is the appropriate status code. Plus, we shouldn't `logger.error`, as that's not a server error - but misuse by the client

Reply:
Fun fact: I checked on a few other servers, and they return a HTTP 200 response with normal content from the non-ws endpoint, totally ignoring that the request was a WS handshake.

Reply:
Yes, checked that with vanilla akka-http and http4s.

Reply:
But I'm convinced we should return HTTP 400 in such case.

Reply:
Yes, this sounds reasonable :)